File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/commands/BookingConfirmationCommands.java
package it.giannibombelli.wsc2026.booking.application.commands;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.primitive.GiftCardReference;
import it.giannibombelli.wsc2026.common.application.Command;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;


public sealed interface BookingConfirmationCommands extends Command<BookingId>
    permits BookingConfirmationCommands.ConfirmBooking, BookingConfirmationCommands.RejectBooking {

    record ConfirmBooking(
        BookingId aggregateId,
        GiftCardReference giftCardReference,
        Money amount
    ) implements BookingConfirmationCommands {
        public ConfirmBooking {
            Require.requireArgument(aggregateId, "bookingId");
            Require.requireArgument(giftCardReference, "giftCardReference");
            Require.requireArgument(amount, "amount");
        }
    }

    record RejectBooking(
        BookingId aggregateId,
        GiftCardReference giftCardReference,
        Money amount
    ) implements BookingConfirmationCommands {
        public RejectBooking {
            Require.requireArgument(aggregateId, "bookingId");
            Require.requireArgument(giftCardReference, "giftCardReference");
            Require.requireArgument(amount, "amount");
        }
    }
}

File: typescript/src/booking/application/commands/bookingConfirmationCommands.ts
// Command per la conferma o il rifiuto di una prenotazione.


import type { Command } from '@/common/application/command.js';
import type { BookingId } from '../../domain/booking/bookingId.js';
import type { GiftCardReference } from '../../domain/primitive/giftCardReference.js';
import type { Money } from '@/common/domain/primitive/money.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type ConfirmBooking = Command<'ConfirmBooking', BookingId> & {
  readonly giftCardReference: GiftCardReference;
  readonly amount: Money;
};

export type RejectBooking = Command<'RejectBooking', BookingId> & {
  readonly giftCardReference: GiftCardReference;
  readonly amount: Money;
};

export type BookingConfirmationCommand = ConfirmBooking | RejectBooking;

export function confirmBooking(aggregateId: BookingId, giftCardReference: GiftCardReference, amount: Money): ConfirmBooking {
  requireArgument(aggregateId, 'bookingId');
  requireArgument(giftCardReference, 'giftCardReference');
  requireArgument(amount, 'amount');
  return { kind: 'ConfirmBooking', aggregateId, giftCardReference, amount };
}

export function rejectBooking(aggregateId: BookingId, giftCardReference: GiftCardReference, amount: Money): RejectBooking {
  requireArgument(aggregateId, 'bookingId');
  requireArgument(giftCardReference, 'giftCardReference');
  requireArgument(amount, 'amount');
  return { kind: 'RejectBooking', aggregateId, giftCardReference, amount };
}
